pydicom/valuerep.py

```diff
@@ -329,15 +330,154 @@ def __reduce_ex__(self, protocol: int) -> Union[str, Tuple[Any, ...]]:
             )
 
 
+def is_valid_ds(s: str) -> bool:
+    """Check whether this string is a valid decimal string.
+
+    Valid decimal strings must be 16 characters or fewer, and contain only
+    characters from a limited set.
+
+    Parameters
+    ----------
+    s: str
+        String to test.
+
+    Returns
+    -------
+    bool
+        True if the string is a valid decimal string. Otherwise False.
+    """
+    # Check that the length is within the limits
+    if len(s) > 16:
+        return False
+
+    # Remove trailing and leading spaces for remainder of the checks
+    s = s.strip()
+
+    # Check all characters are from the allowed subset
+    # part 5, table 6.2-1
+    allowed_chars = {
+        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-', '+',
+        '.', 'e', 'E'
+    }
+    if any(c not in allowed_chars for c in s):
+        return False
+
```

Comment:
I'm a little worried about performance here - for files with huge numbers of DS (e.g. rtstruct) this check over each character of each string could be quite slow.  I'm wondering if some other way could be much faster, either `re` or `str.translate` perhaps, or some use of `isdigit`.  Probably worth taking a bit of time to investigate this now rather than later.

Reply:
I have switched to a regex-based test to determine a valid DS:

```python
re.compile(r'\s*[\+\-]?\d+(\.\d+)?([eE][\+\-]?\d+)?\s*$')
```

I think that covers all cases (it pass all tests and some basic fuzzing but would appreciate a second pair of eyes). The difference in performance is not as big as I might have expected.

Using this test script:

```python                                                                                               
import time
import random

from pydicom.valuerep import is_valid_ds, format_number_as_ds

N = 1000000

test_data = []

# Generate a list of random, correctly formatted DSs
# with good coverage of different exponents
for _ in range(N):
    sign = random.choice([-1, 1])
    base = random.uniform(0, 1)
    exponent = random.randint(-20, 20)
    num = sign * base * 10 ** exponent
    s = format_number_as_ds(num)
    test_data.append(s)


# Measure time taken to validate the full list
start = time.time()
for n in test_data:
    is_valid_ds(n)
end = time.time()

print(end - start)                                                                                   
```

For validation of 1 million strings (`N = 1000000`), I measured 1.597s with the old version (8eb00b9) and 0.499s on the new version with regex (d8f8424) on my i7-7500U laptop and python 3.8.5. So approximately a 3x speedup.

(This also served as a nice fuzzing test of the function and it seemed to work fine in both cases.)